### PR TITLE
Add TableRowClassifier

### LINKED ISSUE #8: Train / predict from tables
It should be pretty straight forward to rip out some parts of the ObjectClassifier and make a new classifier that does not use images, but table as input. 

That would also make this easier to implement:
* #7

### PR DESCRIPTION
**Pull request by kevinyamauchi**
As discussed in #8, this PR adds a classifier for performing classification on rows of a feature table.

Currently, this will accept tables as both pandas DataFrame and a dictionary where the column names are keys and the columns are arrays.

I've added tests that show both table types in `tests/test_table_row_classification.py`. It looks like pytest isn't currently being used, but I think it would be nice to convert the tests in a future PR.

This PR closes #8

### CODE REVIEW

File: tests/test_table_row_classification.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import numpy as np
+import pandas as pd
```
Comment (kevinyamauchi) [inline tests/test_table_row_classification.py:2]: We might want to add a separate set of dependencies for the test suite. Currently, pandas doesn't need to be a dependency for usage of the library (it's used in type annotation, but we can hide that).

File: apoc/_object_classifier.py [hunk 2]
```diff
@@ -180,14 +174,14 @@ def _select_features(self, all_features, features_to_select, labels, ground_trut
 
         Returns
         -------
-        result:list[vector]
+        result:Dict[str, np.ndarray]
             list of vectors corresponding to the requested features. The vectors are shaped (n+1) for n labels. The
             first element corresponds to background.
         ground_truth: ndimage
             selected elements of provided ground truth where it's not 0
```
Comment (kevinyamauchi) [inline apoc/_object_classifier.py:181]: From my testing, I think these vectors are actually length n. Is that correct?

Comment (kevinyamauchi) [reply apoc/_object_classifier.py:181]: (I also need to update the description of the type in the docstring)

File: apoc/_table_row_classifier.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        -------
+        ordered_features : List[np.ndarray]
+            The features stored in a list. The order of the features is
+            specified by self.ordered_feature_names
+        """
+        if len(self._ordered_feature_names) == 0:
+            # if the feature names haven't previously been stored,
+            # store them as a list
+            self._ordered_feature_names = list(feature_table.keys())
+            self.classifier.feature_specification = " ".join(self.ordered_feature_names)
```
Comment (kevinyamauchi) [inline apoc/_table_row_classifier.py:162]: My understanding is that the `feature_specification` is a string of the column names, each one separated by a space. Is that correct?

**Comment by kevinyamauchi** [comment]
Hey, @haesleinhuepf ! I think this is ready for a first round of review.

File: apoc/_object_classifier.py (hunk 2 above)
Comment (haesleinhuepf) [reply apoc/_object_classifier.py:181]: Yes, all looks good! The docstring was wrong.

File: apoc/_table_row_classifier.py (hunk 3 above)
Comment (haesleinhuepf) [reply apoc/_table_row_classifier.py:162]: Yes! For backwards-compatibility, also comma-separated lists should work.

File: apoc/_table_row_classifier.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+import numpy as np
+
+from ._pixel_classifier import PixelClassifier
+
+
+class TableRowClassifier():
+    def __init__(
+            self,
+            opencl_filename="temp_object_classifier.cl",
```
Comment (haesleinhuepf) [inline apoc/_table_row_classifier.py:11]:
```suggestion
            opencl_filename="temp_table_row_classifier.cl",
```

**Review by haesleinhuepf — APPROVED** [review]
Hey @kevinyamauchi 

awesome, thanks so much for working on this! As the tests still pass, I'm expecting it's backwards compatible, wonderful! We should add a test that an old classifier file also works with the new classifier. I'm suspecting this will fail because in the old ObjectClassifier.cl files, the classifier will be specified as "ObjectClassifier" and the new classifier may expect "TableRowClassifier" as classifier type. The fix for this can be simple though. 

Also, would you mind adding a demo notebook showing how the API works? That's one of the first things I typically look at :-)

Thanks again and have a great weekend!

Robert

**Comment by kevinyamauchi** [comment]
> I'm suspecting this will fail because in the old ObjectClassifier.cl files, the classifier will be specified as "ObjectClassifier" and the new classifier may expect "TableRowClassifier" as classifier type. The fix for this can be simple though.

I  actually think this should work. We pass the classifier name when we instantiate the `TableRowClassifier` in the `ObjectClassifier`. Thus, I think this should be fully backwards compatible.

> Also, would you mind adding a demo notebook showing how the API works? That's one of the first things I typically look at :-)

I can add one before we merge.

File: tests/test_table_row_classification.py (hunk 1 above)
Comment (kevinyamauchi) [reply tests/test_table_row_classification.py:2]: I've just added pandas to the github actions workflow for now. We can add a test optional dependencies in a follow up PR if desired.

File: tests/test_table_row_classification.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+feature_table_dict = {
+    "area": np.array([10, 10, 30, 30, 10]),
+    "surface_area": np.array([5, 3, 80, 80, 9])
+}
+feature_table_data_frame = pd.DataFrame(feature_table_dict)
+ground_truth = np.array([1, 1, 2, 2, 1])
+
+
+@pytest.mark.parametrize("feature_table", [feature_table_dict, feature_table_data_frame])
```
Comment (kevinyamauchi) [inline tests/test_table_row_classification.py:13]: @haesleinhuepf , here I used the `parametrize` fixture in pytest to use the same test, but pass in different feature table types. https://docs.pytest.org/en/6.2.x/parametrize.html

File: tests/test_table_row_classification.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+def test_table_row_classification(tmpdir, feature_table):
```
Comment (kevinyamauchi) [inline tests/test_table_row_classification.py:14]: @haesleinhuepf , here I used the `tmpdir` to create a temporary directory in which we write the openCL file. This temporary directory automatically gets cleaned up after the test runs. This is especially nice when running tests locally where you don't want your folder to get polluted with old test files. https://docs.pytest.org/en/6.2.x/tmpdir.html

**Comment by kevinyamauchi** [comment]
Thanks for the review, @haesleinhuepf ! I have added a notebook demonstrating the `TableRowClassifier` and updated the tests. In case you were curious I made comments on places where I "pytest-ified" the tests. I think this is ready for a final review!

**Review by haesleinhuepf — APPROVED** [review]
Awesome, that all looks fantastic @kevinyamauchi ! 

Thanks so much for working on this!

**Pull request merged**